File: WinI2Spider/I2Lib/Quotes/QuoteManager.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace I2Lib.Quotes
{
    public class QuoteManager : Util.JsonListener
    {
        public enum Stage
        {
            sNone,
            sCode,
            sLastDone,
            sDayHigh,
            sDayLow,
            sBestBuy,
            sBestSell,
            sVol,
            sChg,
            sChgP,
            sBuyVol,
            sSellVol,
            sOpenValue,
            sCloseValue
        };

        private Stage mStage = Stage.sNone;
        private int objectCounter;
        private Dictionary<string, Quote> mQuotes = new Dictionary<string, Quote>();
        private Quote mCurrQuote = null;
        private List<QuoteManagerListener> mListeners = new List<QuoteManagerListener>();

        public QuoteManager()
        {
        }

        public void clear()
        {
            mQuotes.Clear();
        }

        public void addListener(QuoteManagerListener listener)
        {
            mListeners.Add(listener);
        }

        public void objectEnd()
        {
            objectCounter--;
            //if(objectCounter == 0)
            //locations.add(curLonLat);
        }

        public void objectStart()
        {
            objectCounter++;
        }

        public void header(string _header)
        {
            string trim_header = Util.JsonParser.trimQuotes(_header);

            if(trim_header == "N")
            {
                mStage=Stage.sCode;
            }
            else if (trim_header == "LT")
            {
                mStage = Stage.sLastDone;
            }
            else if (trim_header == "H")
            {
                mStage = Stage.sDayHigh;
            }
            else if (trim_header == "L")
            {
                mStage = Stage.sDayLow;
            }
            else if (trim_header == "B")
            {
                mStage = Stage.sBestBuy;
            }
            else if (trim_header == "S")
            {
                mStage = Stage.sBestSell;
            }
            else if (trim_header == "VL")
            {
                mStage = Stage.sVol;
            }
            else if (trim_header == "C")
            {
                mStage = Stage.sChg;
            }
            else if (trim_header == "P")
            {
                mStage = Stage.sChgP;
            }
            else if (trim_header == "BV")
            {
                mStage = Stage.sBuyVol;
            }
            else if (trim_header == "SV")
            {
                mStage = Stage.sSellVol;
            }
            else if (trim_header == "O")
            {
                mStage = Stage.sOpenValue;
            }
            else if (trim_header == "PV")
            {
                mStage = Stage.sCloseValue;
            }
            else
            {
                mStage = Stage.sNone;
            }
        }

        public Quote getQuote(string code)
        {
            if (mQuotes.ContainsKey(code))
            {
                return mQuotes[code];
            }
            return null;
        }

        public void data(string _data)
        {
            string trim_data = Util.JsonParser.trimQuotes(_data);
            if (mCurrQuote == null)
            {
                mCurrQuote = new Quote();
            }

            switch (mStage)
            {
                case Stage.sCode:
                    {
                        string ck = mCurrQuote.Code;
                        if (ck == "")
                        {
                            mCurrQuote.Code = trim_data;
                            mQuotes[trim_data] = mCurrQuote;
                        }
                        else if (ck != trim_data)
                        {
                            mCurrQuote = new Quote();
                            mCurrQuote.Code = trim_data;
                            mQuotes[trim_data] = mCurrQuote;
                        }
                    
                        break;
                    }
                case Stage.sLastDone:
                    {
                        mCurrQuote.LastDone = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sDayHigh:
                    {
                        mCurrQuote.DayHigh = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sDayLow:
                    {
                        mCurrQuote.DayLow = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sBestBuy:
                    {
                        mCurrQuote.BestBuy = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sBestSell:
                    {
                        mCurrQuote.BestSell = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sVol:
                    {
                        mCurrQuote.Vol = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sChg:
                    {
                        mCurrQuote.Chg = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sChgP:
                    {
                        mCurrQuote.ChgP = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sBuyVol:
                    {
                        mCurrQuote.BuyVol = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sSellVol:
                    {
                        mCurrQuote.SellVol = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sOpenValue:
                    {
                        mCurrQuote.OpenValue = Convert.ToDouble(trim_data);
                        break;
                    }
                case Stage.sCloseValue:
                    {
                        mCurrQuote.CloseValue = Convert.ToDouble(trim_data);
                        break;
                    }
            }


        }

        public Dictionary<string, Quote> Quotes
        {
            get
            {
                return mQuotes;
            }
        }

        public void arrayEnd()
        {

        }

        public void arrayStart()
        {
        }

        public void parseJson(string json)
        {
            clear();

            Util.JsonParser parser = new Util.JsonParser();
            parser.addListener(this);

            parser.parseString(json);

            if (mCurrQuote != null && mCurrQuote.Code != "")
            {
                mQuotes[mCurrQuote.Code] = mCurrQuote;
                mCurrQuote = new Quote();
            }

            foreach (QuoteManagerListener listener in mListeners)
            {
                listener.onParsed(this);
            }
        }
    }
}

File: I2Web/Version 0/QuoteManager.php
<?php 
require_once("JSONParser.php");
require_once("QuoteInstance.php");

class QuoteManager extends JSONListener
{ 
	// singleton instance 
	private static $instance; 
	private $parser;
	private $mQuotes;
	private $mStage;
	private $breakpoint_reached;
	private $mTargetCode;
	private $mTargetQuote;

	// private constructor function 
	// to prevent external instantiation 
	private function __construct() 
	{ 
		$this->parser=new JSONParser();
		$this->parser->addListener($this);
		$this->mQuotes=array();
		$this->mStage='sNone';
		$this->breakpoint_reached=false;
		$this->mTargetCode='';
		$this->mTargetQuote=new QuoteInstance();
	}

	// getInstance method 
	public static function getInstance() 
	{ 
		if(!self::$instance) 
		{ 
		  self::$instance = new self(); 
		} 

		return self::$instance; 
	}

	public function header($_header)
	{
		$header_content=$this->trimQuotes($_header);
		//echo $header_content . "<br />";

		if($header_content == "N")
		{
			$this->mStage='sCode';
		}
		else if($header_content == "LT") 
		{
			$this->mStage = 'sLastDone';
		}
		else if($header_content == "H") 
		{
			$this->mStage = 'sDayHigh';
		}
		else if($header_content == "L") 
		{
			$this->mStage = 'sDayLow';
		}
		else if($header_content == "B") 
		{
			$this->mStage='sBestBuy';
		}
		else if($header_content == "S") 
		{
			$this->mStage='sBestSell';
		}
		else if($header_content == "VL") 
		{
			$this->mStage ='sVol';
		}
		else if($header_content == "C") 
		{
			$this->mStage='sChg';
		}
		else if($header_content == "P") 
		{
			$this->mStage='sChgP';
		}
		else if($header_content == "BV") 
		{
			$this->mStage='sBuyVol';
		}
		else if($header_content == "SV") 
		{
			$this->mStage='sSellVol';
		}
		else if($header_content == "O") 
		{
			$this->mStage='sOpenValue';
		}
		else if($header_content == "PV") 
		{
			$this->mStage='sCloseValue';
		}
		else
		{
			$this->mStage = 'sNone';
		}
		
		
		return false;
		
	}

	public function data($_value)
	{
		$_value=$this->trimQuotes($_value);
		if($_value==$this->mTargetCode && $this->mStage=='sCode')
		{
			$this->breakpoint_reached=true;
		}
		if($this->breakpoint_reached==true)
		{
			if($this->mStage=='sCode')
			{
				$this->mTargetQuote->setCode($_value);
			}
			else if($this->mStage=='sLastDone')
			{ 
				$this->mTargetQuote->setLastDone((float)($_value));
			}
			else if($this->mStage=='sDayHigh')
			{ 
				$this->mTargetQuote->setDayHigh((float)($_value));
			}
			else if($this->mStage=='sDayLow')
			{ 
				$this->mTargetQuote->setDayLow((float)($_value));
			}
			else if($this->mStage=='sBestBuy')
			{ 
				$this->mTargetQuote->setBestBuy((float)($_value));
			}
			else if($this->mStage=='sBestSell')
			{ 
				$this->mTargetQuote->setBestSell((float)($_value));
			}
			else if($this->mStage=='sVol')
			{ 
				$this->mTargetQuote->setVol((float)($_value));
			}
			else if($this->mStage=='sChg')
			{ 
				$this->mTargetQuote->setChg((float)($_value));
			}
			else if($this->mStage=='sChgP')
			{ 
				$this->mTargetQuote->setChgP((float)($_value));
			}
			else if($this->mStage=='sBuyVol')
			{ 
				$this->mTargetQuote->setBuyVol((float)($_value));
			}
			else if($this->mStage=='sSellVol')
			{ 
				$this->mTargetQuote->setSellVol((float)($_value));
			}
			else if($this->mStage=='sOpenValue')
			{ 
				$this->mTargetQuote->setOpenValue((float)($_value));
			}
			else if($this->mStage=='sCloseValue')
			{ 
				$this->mTargetQuote->setCloseValue((float)($_value));
			}
		}
		return false;
	}
	
	public function trimQuotes($text)
	{
		$len=strlen($text);
		if($len == 0) return "";
		if($text[0]=='"' && $text[$$len-1]=='"')
		{
			return substr($text, 1, $len-2);
		}
		else if($text[0]=='"')
		{
			return substr($text, 1, $len-1);
		}
		if($text[0]=='\'' && $text[$len-1]=='\'')
		{
			return substr($text, 1, $len-2);
		}
		else if($text[0]=='\'')
		{
			return substr($text, 1, $len-1);
		}
		return substr($text, 0, $len);
	}

	public function arrayStart()
	{
		return false;
	}

	public function arrayEnd()
	{
		return false;
	}

	public function objectStart()
	{
		return false;
	}

	public function objectEnd()
	{
		if($this->breakpoint_reached==true)
		{
			return true;
		}
		return false;
	}
	
	public function load($url)
	{
		$this->breakpoint_reached=false;
		
		/*
		$content = '';
		if ($fp = fopen($url, 'r'))
		{
			// keep reading until there's nothing left 
			while ($line = fread($fp, 1024)) 
			{
			  $content .= $line;
			}

			//$content=str_replace("{}&&", "", $content);
			//echo $content;
			$this->parser->parseString($content);
			
		}
		*/
		
		/*
		$content=file_get_contents($url);
		$this->parser->parseString($content);
		*/
		
		
		$ch = curl_init($url);
		$fp = @fopen("temp.txt", "w");
		curl_setopt($ch, CURLOPT_FILE, $fp);
		curl_setopt($ch, CURLOPT_HEADER, 0);
		curl_exec($ch);
		curl_close($ch);
		fclose($fp);
		$file = "temp.txt";
		if ($fp = fopen($file, 'r'))
		{
			// keep reading until there's nothing left 
			while ($line = fread($fp, 1024)) 
			{
			  $content .= $line;
			}

			//$content=str_replace("{}&&", "", $content);
			//echo $content;
			$this->parser->parseString($content);
			
		}
		
	}
	
	public function find($url, $targetCode)
	{
		$this->mTargetCode=$targetCode;
		$this->load($url);
		return $this->getTargetQuote();
	}
	
	public function clear()
	{
		unset($this->mQuotes);
		$this->mQuotes=array();
	}
	
	public function getTargetQuote()
	{
		return $this->mTargetQuote;
	}
} 
?>